lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -175,19 +191,74 @@ data SelectionLimit
 
 -- | The result of performing a successful selection.
 --
-data SelectionResult = SelectionResult
+data SelectionResult change = SelectionResult
```

Comment:
I can see why a parameterized type is tempting here (as it's convenient), but I can't help wondering if there isn't a more straightforward solution.

If I've understood correctly, `SelectionResult` can be instantiated in two ways, for two different use cases:

1. **`SelectionResult TokenBundle`:**
      the result of running `performSelection`
      
2. **`SelectionResult TxOut`:**
      a balanced selection where change addresses have been assigned

Unless I'm mistaken, the second use case is not required by the `performSelection` function, and is only really used outside of the `RoundRobin` module.

## Proposal

What if we were to instead just make a new record type for the second use case, like this:

```hs
--- | A balanced coin selection where all change addresses have been assigned.
data Selection = Selection
    { inputs  :: ![(TxIn, TxOut)]
    , outputs :: ![TxOut]
    , change  :: ![TxOut]
    }
```

We'd produce a `Selection` by post-processing a `SelectionResult` and combining it with the original criteria. Something like this:
```hs
makeSelection
    :: GenChange s
    => SelectionCriteria
    -> SelectionResult
    -> s Selection
```

Then, unless I've misunderstood, we wouldn't have to include `outputsSelected` in the `SelectionResult` type, which removes the need to write a test that `performSelection` returns the correct value of `outputsSelected` (I think).

And then `mkTransaction` would have the following type:
```hs
        -> ProtocolParameters
            -- Current protocol parameters
        -> TransactionCtx
            -- An additional context about the transaction
        -> Selection
            -- A balanced coin selection where all change addresses have been
            -- assigned.
```

Thoughts?

Reply:
We've got already quite many types all over the place when it comes to selections (and there are more in the transaction module like `UnsignedTx`). Thus, instead of introducing yet-a-new-type which has almost exactly the same semantic as an existing one but only differs on one particular field, I found it preferable to parameterize over the change type. 

Not only does it allow the reader to keep the same context when processing function using `SelectionResult TxOut` or `SelectionResult TokenBundle` (it's quite clear that it's roughly the same thing), but it also allow for re-using functions on that type quite easily. Plus, It makes type-signature quite clear about what's happening:

```hs
assignChangeAddresses
     :: SelectionResult TokenBundle
    -> SelectionResult TxOut
```

Neat.

Reply:
@KtorZ I can definitely sympathize with your argument!

But I also think there are drawbacks to having a function return part of its input within its output.

An unfamiliar reader of `performSelection` might very well wonder:
- does `outputsCovered` really have the same contents as `outputsToCover`?
- is it a subset, or perhaps something else?

Before this PR, there could be no such questions in the reader's mind, because that field didn't exist. But after applying this PR, the reader will have to examine the definition of `performSelection` to confirm the answer.

To fix this, I think we would also need to:

1. Upgrade our property tests for `performSelection` to ensure that the intended relationship between `SelectionCriteria.outputsToCover` and `SelectionResult.outputsCovered` really does hold.

2. Describe the intended relationship between `SelectionCriteria.outputsToCover` and `SelectionResult.outputsCovered` in the documentation for `performSelection`, so that it's clear to the reader.

Reply:
@KtorZ I've created a PR (https://github.com/input-output-hk/cardano-wallet/pull/2469) which adds:
 
> 1. Upgrade our property tests for `performSelection` to ensure that the intended relationship between `SelectionCriteria.outputsToCover` and `SelectionResult.outputsCovered` really does hold.
> 2. Describe the intended relationship between `SelectionCriteria.outputsToCover` and `SelectionResult.outputsCovered` in the documentation for `performSelection`, so that it's clear to the reader.

I still believe this solution to be worse than adding a new type (and leaving `performSelection` alone), but if we have to choose it, I think we should have tests.

Reply:
> Before this PR, there could be no such questions in the reader's mind, because that field didn't exist. But after applying this PR, the reader will have to examine the definition of performSelection to confirm the answer.
> 
> I think this is non-ideal, and would prefer to keep the range of performSelection as minimal as possible

It seems like you're not objecting the parameterization so much but actually object the additions of the few other fields to the result. However, these are *necessary* for the rest of the code to work and to integrate with the new selection algorithms. Without that, we'd have to duplicate logic from `performSelection` from the same argument to re-construct the augmented result. 

This is generally not advised and has already caused us issues in the past. This PR actually resolves a lot of the logic duplication issues our current coin selection integration has and the best way to achieve this at this stage is to make sure that all information needed are returned by `performSelection`.

Reply:
> It seems like you're not objecting the parameterization so much but actually object the additions of the few other fields to the result. 

That is indeed my main objection. I think the `performSelection` function is already quite complicated, and that the combination of inputs with outputs could be handled by another function that calls `performSelection`, returning some kind of record that combines `SelectionCriteria` with `SelectionResult`. This way, we wouldn't have to alter the definition of `performSelection`, or extend any of its property tests.

However, given our time constraints, I do think we can go with the solution in this PR, but with some extra small additions similar to:

- e628436fe88cd8905a557eebfcc921f987c82c6d
- 5e3d602ca4b8a0121a4b36b9e5ab3b766db7f497

(Feel free to cherry pick if you think these are useful.)